embedded-graphics/src/primitives/rounded_rectangle.rs

```diff
@@ -0,0 +1,777 @@
+//! The rounded rectangle primitive.
+
+use crate::{
+    drawable::{Drawable, Pixel},
+    geometry::{Dimensions, Point, Size},
+    pixelcolor::PixelColor,
+    primitives::{
+        ellipse_quadrant::{self, EllipseQuadrant, Quadrant},
+        rectangle::{self, Rectangle},
+        ContainsPoint, Primitive,
+    },
+    style::{PrimitiveStyle, Styled},
+    transform::Transform,
+    DrawTarget,
+};
+
+/// The definition of each corner radius for a rounded rectangle.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct CornerRadii {
+    /// Top left corner radius
+    pub top_left: Size,
+
+    /// Top right corner radius
+    pub top_right: Size,
+
+    /// Bottom right corner radius
+    pub bottom_right: Size,
+
+    /// Bottom left corner radius
+    pub bottom_left: Size,
+}
+
+impl CornerRadii {
+    /// Create a new set of corner radii with all corners having equal values.
+    pub fn new_equal(radius: Size) -> Self {
+        Self {
+            top_left: radius,
+            top_right: radius,
+            bottom_right: radius,
+            bottom_left: radius,
+        }
+    }
+
+    /// Confine corner radii that are too large to a given bounding rectangle
+    fn confine(self, bounding_box: Size) -> Self {
+        // Compute free space (linear region) length along each edge. If there is no free space
+        // left, the value will be negative.
+        let free_space_top =
+            bounding_box.width as i32 - self.top_left.width as i32 - self.top_right.width as i32;
+        let free_space_right = bounding_box.height as i32
+            - self.top_right.height as i32
+            - self.bottom_right.height as i32;
+        let free_space_bottom = bounding_box.width as i32
+            - self.bottom_left.width as i32
+            - self.bottom_right.width as i32;
+        let free_space_left = bounding_box.height as i32
+            - self.top_left.height as i32
+            - self.bottom_left.height as i32;
```

Comment:
Thoughts on how to strip all the `as i32`s? I considered adding `TryFrom<Size> for Point` which would probably make this cleaner, but would either require `Result` or some `unwrap()`s which I'm not keen on.

Reply:
You could use `checked_sub`:
```rust
let overlap_top = (self.top_left.width + self.top_right.width).checked_sub(bounding_box.width);
```
This would return `None` if there is no overlap and `Some(u32)` if it overlaps.